File: tree/leetcode_589_n-ary-tree-preorder-traversal.cpp
/**
 * 589. N叉树的前序遍历
 * https://leetcode-cn.com/problems/n-ary-tree-preorder-traversal/
 */

/*
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/
class Solution {
public:
    // 递归
    // 时间复杂度：O(n)
    // 空间复杂度：O(n)
    vector<int> preorder1(Node* root) {
        vector<int> res;
        traverse(root, res);
        return res;
    }

    void traverse(Node* node, vector<int>& res){
        if(node != NULL){ 
            res.push_back(node->val);
            
            for(Node* child : node->children){
                traverse(child, res);
            }
        }
    }

    // 迭代
    // 时间复杂度：O(n)
    // 空间复杂度：O(n)
    vector<int> preorder(Node* root) {
        vector<int> res;
        if(root == nullptr){
            return res;
        }
        
        stack<Node *> stk;
        stk.push(root);
        while(!stk.empty()){
            Node* curNode = stk.top();
            stk.pop();
            res.push_back(curNode->val);
            
            for(int i = curNode->children.size()-1; i >= 0; i--){
                if(curNode->children[i] != nullptr){
                    stk.push(curNode->children[i]);
                }
            }
        }
        
        return res;
    }
};

File: tree/leetcode_589_n-ary-tree-preorder-traversal.py
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""
class Solution(object):
    # 利用栈
    # 时间复杂度：O(n)
    # 空间复杂度：O(n)
    def preorder(self, root):
        """
        :type root: Node
        :rtype: List[int]
        """
        if root is None:  # 此处如果使用 not root，则执行时间会下降
            return []

        stack, output = [root, ], []
        while stack:
            root = stack.pop()
            output.append(root.val)
            stack.extend(root.children[::-1])

        return output